pkg/signer/signer.go

```diff
@@ -0,0 +1,269 @@
+// to authenticate that a given release image digest has been signed by a trusted party.
+type Interface interface {
+	Verify(ctx context.Context, releaseDigest, location string, signature []byte) error
+	Sign(releaseDigest, pullSpec string) ([]byte, error)
+}
+
+func loadArmoredOrUnarmoredGPGKeyRing(data []byte) (openpgp.EntityList, error) {
+	keyring, err := openpgp.ReadArmoredKeyRing(bytes.NewReader(data))
+	if err == nil {
+		return keyring, nil
+	}
+	return openpgp.ReadKeyRing(bytes.NewReader(data))
+}
+
+type releaseSigner struct {
+	signer    *openpgp.Entity
+	verifiers map[string]openpgp.EntityList
+}
+
+func NewFromKeyring(path string) (Interface, error) {
+	data, err := ioutil.ReadFile(path)
+	if err != nil {
+		return nil, err
+	}
+	keyring, err := loadArmoredOrUnarmoredGPGKeyRing(data)
+	if err != nil {
+		return nil, err
+	}
+	var signer *openpgp.Entity
+	for _, key := range keyring {
+		if key.PrimaryKey.CanSign() && key.PrivateKey != nil {
+			signer = key
+			break
+		}
+	}
+	if signer == nil {
+		return nil, fmt.Errorf("the provided keyring must contain a private key capable of signing")
+	}
+	name := filepath.Base(path)
+	name = strings.TrimSuffix(name, filepath.Ext(name))
+	return &releaseSigner{
+		signer: signer,
+		verifiers: map[string]openpgp.EntityList{
+			name: keyring,
+		},
+	}, nil
+}
+
+// String summarizes the verifier for human consumption
+func (s *releaseSigner) String() string {
+	var keys []string
+	for name := range s.verifiers {
+		keys = append(keys, name)
+	}
+	sort.Strings(keys)
+
+	var builder strings.Builder
+	builder.Grow(256)
+	fmt.Fprintf(&builder, "All release image digests must have GPG signatures from")
+	if len(keys) == 0 {
+		fmt.Fprint(&builder, " <ERROR: no verifiers>")
+	}
+	for _, name := range keys {
+		verifier := s.verifiers[name]
+		fmt.Fprintf(&builder, " %s (", name)
+		for i, entity := range verifier {
+			if i != 0 {
+				fmt.Fprint(&builder, ", ")
+			}
+			if entity.PrimaryKey != nil {
+				fmt.Fprintf(&builder, strings.ToUpper(fmt.Sprintf("%x", entity.PrimaryKey.Fingerprint)))
+				fmt.Fprint(&builder, ": ")
+			}
+			count := 0
+			for identityName := range entity.Identities {
+				if count != 0 {
+					fmt.Fprint(&builder, ", ")
+				}
+				fmt.Fprintf(&builder, "%s", identityName)
+				count++
+			}
+		}
+		fmt.Fprint(&builder, ")")
+	}
+	return builder.String()
+}
+
+func (s *releaseSigner) Sign(releaseDigest, pullSpec string) ([]byte, error) {
+	if len(releaseDigest) == 0 || len(pullSpec) == 0 {
+		return nil, fmt.Errorf("you must specify a release digest and pull spec to sign")
+	}
+
+	sig := &signature{
+		Critical: criticalSignature{
+			Type: "atomic container signature",
+			Image: criticalImage{
+				DockerManifestDigest: releaseDigest,
+			},
+			Identity: criticalIdentity{
+				DockerReference: pullSpec,
+			},
+		},
+		Optional: optionalSignature{
+			Creator:   "openshift release-controller",
```

Comment:
It would be useful to have a version number, or something similar, in the value. (So that future code can detect and work-around possible bugs in signatures created by earlier versions.)